pynumaflow/sink/async_sink.py

```diff
@@ -0,0 +1,198 @@
+
+
+class AsyncSink(udsink_pb2_grpc.UserDefinedSinkServicer):
+    """
+    Provides an interface to write a User Defined Sink (UDSink)
+    which will be exposed over gRPC.
+
+    Args:
+        sink_handler: Function callable following the type signature of UDSinkCallable
+        sock_path: Path to the UNIX Domain Socket
+        max_message_size: The max message size in bytes the server can receive and send
+        max_threads: The max number of threads to be spawned;
+                     defaults to number of processors x 4
+
+    Example invocation:
+    >>> from typing import List
+    >>> from pynumaflow.sink import Datum, Responses, Response, Sink
+    >>> def my_handler(datums: Iterator[Datum]) -> Responses:
+    ...   responses = Responses()
+    ...   for msg in datums:
+    ...     responses.append(Response.as_success(msg.id))
+    ...   return responses
+    >>> grpc_server = Sink(my_handler)
+    >>> grpc_server.start()
+    """
+
+    def __init__(
+            self,
+            sink_handler: UDSinkCallable,
+            sock_path=SINK_SOCK_PATH,
+            max_message_size=MAX_MESSAGE_SIZE,
+            max_threads=MAX_THREADS,
+    ):
+        self.background_tasks = set()
+        self.__sink_handler: UDSinkCallable = sink_handler
+        self.sock_path = f"unix://{sock_path}"
+        self._max_message_size = max_message_size
+        self._max_threads = max_threads
+        self.cleanup_coroutines = []
+
+        self._server_options = [
+            ("grpc.max_send_message_length", self._max_message_size),
+            ("grpc.max_receive_message_length", self._max_message_size),
+        ]
+
+    async def SinkFn(
+            self, request_iterator: AsyncIterable[udsink_pb2.DatumRequest], context: NumaflowServicerContext
+    ) -> udsink_pb2.ResponseList:
+        """
+        Applies a sink function to a list of datum elements.
+        The pascal case function name comes from the proto udsink_pb2_grpc.py file.
+        """
+        # if there is an exception, we will mark all the responses as a failure
+        datum_iterator = datum_generator(request_iterator=request_iterator)
+        response_task = asyncio.create_task(
+            self.__async_sink_handler(datum_iterator)
+        )
+
+        # Save a reference to the result of this function, to avoid a
+        # task disappearing mid-execution.
+        self.background_tasks.add(response_task)
+        response_task.add_done_callback(lambda t: self.background_tasks.remove(t))
+
+        await response_task
+        results_futures = response_task.result()
+        resp =[]
+
+        for fut in results_futures:
+            await fut
+            resp.append(fut.result())
+
+        return udsink_pb2.ResponseList(responses=resp)
+
+    async def __async_sink_handler(self, datum_iterator: AsyncIterable[Datum]):
+        callable_dict = {}
+        # iterate through all the values
+        async for d in datum_iterator:
+            keys = d.keys()
```

Comment:
we dont need all of this. sink does not need to track keys and so we can use the same iterator(decorated) and pass it along like we discussed.

Reply:
Done